using ConanCenter in prod. TutorialDevops (#3322)
conan-io/docs: devops/consuming_conan_center.rst

.. _devops_consuming_conan_center:

Using ConanCenter packages in production
========================================

ConanCenter is a fantastic resource that contains a huge knowledge base and a reference implementation of recipes contributed by the community on how to create and build Conan packages for a lot of open source third party libraries and applications supporting multiple platforms.

It also provides pre-compiled binaries for a wide range of configurations: multiple OS, including Windows, Linux and OSX, a variety of compilers, compiler versions, static and shared libraries, etc. These precompiled packages can be very useful to do quick experiments, testing, onboarding, and they also serve as a proof that the recipes for those packages can work solidly for that wide range of configurations.

> However, in the same way modern Devops discourages to depend directly on any central repository, it is not advised to depend directly on ConanCenter for production environments, for several reasons:

Reviewer: Trying to squeeze some extra emphasis 

```suggestion
However, in the same way modern DevOps discourages to depend directly on any central repository, **it is not advised to depend directly on ConanCenter for production environments**, for several reasons:
```

- ConanCenter accepts contributions that require the most recently released versions of Conan. If you are in an environment where you are unable to keep Conan up-to-date across all your systems (developers, CI, etc) - depending directly on recipes from ConanCenter carries the risk of retrieving recipes that do not work with your older version of Conan.
- The changes in recipes might also unpexectedly cause bugs and other issues. Recipes and package updates might bump dependencies versions to use the latest, forcing a continuous update over all dependencies versions. That can cause version conflicts if your project doesn't bump dependencies versions at the same pace.

> - Reviews in ConanCenter can take a while to be reviewed and merged, as there is usually a high volume of requests. Please consider this if you may find yourself in a situation where you may need proposed changes merged with urgency. Pull Requests are considered according to priorities and resources, and will only be accepted with positive peer reviews by the Conan team.

Reviewer: I would avoid pointing at review times. 🙊 

I would make this more generic

> If you are working with tight dead lines or on a schedule, it's far more advantageous to control when the changes you require are rolled out.

Just for the top of my head

- Binary configurations can be discontinued in ConanCenter, when new compiler versions are released and ConanCenter starts to build them, it can drop some of the older versions.

> - Outages, maintenance windows and other service interruptions can happen, ranging from a few minutes to days. The infrastructure to host central repositories is typically large and complex, and relies on cloud providers that sometimes have their own issues, networks, CDN, etc.

Reviewer: I think less is more, I don't think we need to enumerate problems or make any suggestions about how long issues can take

```suggestion
- Outages, maintenance windows, and service interruptions can happen. The infrastructure to host central repositories is typically large and complex, and relies on cloud providers.
```

- Supply chain issues. Even if the best effort is made to make ConanCenter secure and the process to add packages to ConanCenter is one of the most strict ones with human reviews for every package and many automated checks and processes, eventually all systems and central repositories in every technology are subject to attempts of supply chain attacks.

Most of these reasons are not unique to Conan, but to every other package manager for other programming languages. 

.. important::
  

>   The general devops known good practice is that you shouldn't rely on central repositories on production systems, but do consider the following approaches:

Reviewer: My brain being special, it's not spelled the same way everywhere 

```suggestion
  The general DevOps known good practice is that you shouldn't rely on central repositories on production systems, but do consider the following approaches:
```

Reviewer: There was a double negative in the sentence. I tired to word it more clearly. 

```suggestion
  A known good practice with DevOps is to reduce reliance on central repositories for production systems. Consider the following approaches:
```

>   - If your project is mostly a "consuming" project, i.e. you are not generating your own packages, but just consuming packages from ConanCenter, and the above issues are not a concern because your project does not need that kind of robustness or security, then you can use **lockfiles** to at least guarantee that your project build using the same dependencies over time. Check :ref:`the lockfile docs<tutorial_versioning_lockfiles>` for more information.

Reviewer: Picking the word that match the CLI so there's no ambiguity 

```suggestion
  - If your project is mostly a "consuming" project, i.e. you are not creating or uploading your own packages, but just installing packages from ConanCenter, and the above issues are not a concern because your project does not need that kind of robustness or security, then you can use **lockfiles** to at least guarantee that your project build using the same dependencies over time. Check :ref:`the lockfile docs<tutorial_versioning_lockfiles>` for more information.
```

--------------------------------------------------

Hosting your own copy of the packages you need in your server could be done by just downloading binaries from ConanCenter and then uploading them to your own server. However, it is much better to fully own the complete supply chain and create the binaries in your own CI systems. So the recommended flow to use ConanCenter packages in production would be:

- Create a fork of the ConanCenter Github repository: https://github.com/conan-io/conan-center-index
- Create a list of the packages and versions you need for your projects. This list can be added to the fork too, and maintained there (packages can be added and removed with PRs when the teams need them).
- Create a script that first ``conan export`` all the packages in your list, then ``conan create --build=missing`` them. Do not add ``user/channel`` to these packages, it is way simpler to use them as ``zlib/1.2.13`` without user-channel. The ``user/channel`` part would be mostly recommended for your own proprietary packages, but not for open source ConanCenter packages.
- Upload your build packages to your own server, that you use in production, instead of ConanCenter.

> This is the basic flow idea. We will be adding examples and tools to further automate this flow as soon as possible.

Reviewer: It would be amazing to have blank pages as place holders so people searching for this can find them

Just link back here in the meantime so the information is surfaced 🙏 

The docs search matches page titles and so it would be easier to find

This flow is relatively straightforward, and has many advantages that mitigate the above risks:

- No central repository outage can affect your builds.

> - No changes in the central repository can break your projects, you are in full control when and how those changes are updated in your packages (as explained below).

Reviewer: The last few words. It's not clear which section below explains this

> - You can customize, adapt, fix and perfectly control what versions are used, and release fixes in minutes, not weeks. You can apply customizations that wouldn't be accepted in the central repository.

Reviewer: This is referring to the lows of CCI which I do not think are representative of our goals and trajectory 

I would focus on the benefits to the user :)

```suggestion
- You can customize, adapt, fix and control what versions are used and release fixes on your own schedule as you see fit.
```

> - You fully control the binaries supply chain, from the source (recipes) to the binaries, eliminating in practice the majority of potential supply chain attacks of central repositories.

Reviewer: The second half of the sentence was a bit confusing.

Hopefully this captures the same idea

```suggestion
- You control the binaries supply chain, from the recipes to the binaries, eliminating in practice the majority of potential supply chain attacks seen in central repositories.
```

Updating from upstream
++++++++++++++++++++++

Updating from the upstream ``conan-center-index`` Github repo is still possible, and it can be done in a fully controlled way:

- Merge the latest changes in the upstream main fork of ``conan-center-index`` into your fork.
- You can check and audit those changes if you want to, analyzing the diffs (some automation that trims the diffs of recipes that you don't use could be useful)

> - Firing the above process will efficiently rebuild the new binaries that are needed. If your recipes are not affected by changes, the process will avoid rebuilding binaries (thanks to ``--build=missing``).

Reviewer: Is the process above the section "Creating and hosting your own ConanCenter binaries" it's not clear what this is referring too

- You can upload the packages to a secondary "test" server repository. Then test your project against that test server, to check that your project is not broken by the new ConanCenter packages.
- Once you verify that everything is good with the new packages, you can copy them from the secondary "test" repository to your main production repository to start using them.

> .. note::

Reviewer: I really like how this is very clear and concise... I know this is usually at the bottom of pages else where 

But given the importance I would love for this to be at the top as a TDLR
